### ISSUE TITLE
Unable to access internal storage

### ISSUE THREAD

**Issue by efenex** [xvortex/TWRP_Rai_Zin_32#1]
Hi, I seem to have soft-bricked my Onkyo trying to get this thing to work. Main issue is the following:

<6>[    1.967271] mmc0: no vmmc regulator found
<6>[    1.969640] mmc0: SDHCI controller on msm_sdcc.1 [msm_sdcc.1] using ADMA
<3>[    2.071603] mmc0: unrecognised EXT_CSD revision 8
<3>[    2.071619] mmc0: error -22 whilst initialising MMC card

This is an excerpt of dmesg after booting into TWRP. The end-result is that no block devices are being recognized, and I don't have a /dev/block/platform folder (just loopX and ramX devices). 

Before I figured out what the issue was with TWRP, I realized that I didn't have access to my block devices (but not why), so I thought: hey, I have fastboot, why not just flash the system.img that way? So I tried it, and fastboot started to flash, after spewing but seemingly ignoring a warning about sparse file size and a max of 512MB files - it seemed to chunk the flash into multiple pieces so I still had some hope. However, after flashing the first chunk it crashed (some assertion error), and now not only is my recovery broken, but the DAP no longer boots in Android.

Now, this still looks like a soft-brick to me: I can access fastboot, I can boot into TWRP (or the original recovery using fastboot boot), and I can flash other recoveries, boot - and maybe even system (if i can get one in the right format and/or smaller size). I just can't get my internal block storage to be recognized..

To be clear: TWRP was unable to see my partitions/block storage even *before* I did the fastboot flash system command. I did not properly check the logs so I can't be sure it was the same error as pasted above.

Either way, seems to me my best option would be to get TWRP to recognized the block storage again and after that reflash a stock image (hopefully I didn't lose any partition table details).

Any recommendations or help you could provide would be extremely welcome.

**Comment by efenex** [comment]
FYI, output from fastboot:

# fastboot flash system system.img 
target reported max download size of 536870912 bytes
Invalid sparse file format at header magi
erasing 'system'...
OKAY [  0.013s]
sending sparse 'system' 1/2 (515398 KB)...
OKAY [ 16.809s]
writing 'system' 1/2...
FAILED (remote: Unknown chunk type)
finished. total time: 16.832s



I can still erase/format the system disk just fine from fastboot (but it doesn't help in getting the sdcard recognized)

[root@ 2016-09-02--12-49-57_LMY47V]# fastboot erase system
******** Did you mean to fastboot format this ext4 partition?
erasing 'system'...
OKAY [  0.020s]
finished. total time: 0.020s
[root@ 2016-09-02--12-49-57_LMY47V]# fastboot format:ext4 system
Creating filesystem with parameters:
    Size: 1342177280
    Block size: 4096
    Blocks per group: 32768
    Inodes per group: 8192
    Inode size: 256
    Journal blocks: 5120
    Label: 
    Blocks: 327680
    Block groups: 10
    Reserved block group size: 79
Created filesystem with 11/81920 inodes and 10749/327680 blocks
target reported max download size of 536870912 bytes
erasing 'system'...
OKAY [  0.013s]
sending 'system' (22984 KB)...
OKAY [  0.734s]
writing 'system'...
OKAY [  0.341s]
finished. total time: 1.088s

**Comment by efenex** [comment]
minor progress, but not really:

]# img2simg system.img system-sparse.img
]# file system-sparse.img 
system-sparse.img: Android sparse image, version: 1.0, Total of 327680 4096-byte output blocks in 644 input chunks.
]# fastboot flash system system-sparse.img 
target reported max download size of 536870912 bytes
erasing 'system'...
OKAY [  0.014s]
sending sparse 'system' 1/2 (515398 KB)...
OKAY [ 17.254s]
writing 'system' 1/2...
FAILED (remote: Unknown chunk type)
finished. total time: 17.278s

(the warnings are gone, but it still fails the flash)

**Comment by efenex** [comment]
Trying to build a new version (or same version even) of this recovery: I've managed to get a recovery.img built, but I had to up the recovery partition size in the boardconfig (which means I must be doing something different from you). The recovery image doesn't boot, so I'm doing something wrong for sure. Going to try anther build based on android-5.1 branch in Omni next, as opposed to android-6.0.

**Comment by efenex** [comment]
Another minor update: using the android-5.1 branch to build the TWRP recovery, I do not have to modify the recovery partition size, and I can actually get TWRP to boot. Shortly after booting it appears to freeze/shutdown (before any touch is registered, but after the TWRP splash screen). Have to try and get some debugging active to figure out what happens.

**Comment by efenex** [comment]
Seems like adb is not started yet, as I am unable to logcat while TWRP boots. I tried building without SElinux but this had no effect. Not sure how to get logdata out without logcat for now.

**Comment by efenex** [comment]
Maybe I need an UART to 3.5mm cable to get kernel boot output and debug why this thing won't boot properly?

**Comment by efenex** [comment]
Update, but no progress: 

- I tried booting the original TWRP image with every sdhci.debug_quirks value, as well as every sdhci.debug_quirks2 value. They had some minor effect on the SD-card detection, but nothing useful/productive. I didn't try every possible combination of both settings, because that would take infinitely longer and is unlikely to have any benefit.

- I tried reducing the system.img size to something below < 512MB, as this is the limit that fastboot tells us when we try to flash system. Reducing the image was not a big problem (simg2img, mount, wipe the apps folder, write file full of 0's, remove file, umount, img2simg), but the end-result is sadly the same as ever:

```
fastboot flash system system-sparse.img
target reported max download size of 536870912 bytes
erasing 'system'...
OKAY [  0.013s]
sending 'system' (464065 KB)...
OKAY [ 14.777s]
writing 'system'...
FAILED (remote: Unknown chunk type)
finished. total time: 14.802s
```

- I looked into the kernel used in this TWRP image - it is the exact same one as the kernel of the original stock recovery image and stock boot.img, md5sum:

```
92b9cba72c345817ccf76b4bc676177c  kernel-stockrecovery
92b9cba72c345817ccf76b4bc676177c  kernel-twrp
92b9cba72c345817ccf76b4bc676177c  kernel-stockbootimg
```

In other words, even if I do get a custom TWRP built, it would still have to use the same kernel, and is thus unlikely to provide any difference in terms of access to the eMMC hardware. Unless someone has a newer kernel that is used with this device somewhere...

Kernel sources do not appear to be made available, as per following post by XVortex:

[OnkyoUSA Forum](http://forums.onkyousa.com/viewtopic.php?p=35527&sid=d556b47341d3f0ffd2d45e75b6a0cf4e)

So it doesn't seem likely that I can build a new kernel myself.

**Comment by efenex** [comment]
Update & resolution!

Using the kernel from update 0.1.37.8_20161104.013642 makes my sd-card accessible again. Specifically, the kernel with md5 9aea8b1c2fcd9c255c9ecb633dba4972.

Since I am still unable to compile a new TWRP based on these sources, I picked the original recovery apart and rebuilt it with the other kernel, as follows:

```
bootimgtool -x twrp-3.0.2-0-Rai_Zin_32.img
rm zImage
cp 0.1.37.8_20161104.013642/zImage ./
wget https://github.com/xvortex/TWRP_Rai_Zin_32/raw/master/dt.img
bootimgtool -p parameters.cfg -k zImage -r ramdisk.img -d dt.img -v qcom -c recovery.img
fastboot boot recovery.img
```

I will upload this new recovery image somewhere, and maybe do a PR here.

After getting the storage recognized, the rest of the process was standard; format all the partitions (to fix any damage I may have done using fastboot flash), flash a system zip and boot.

Hopefully this will help someone in the future, I sure spent a lot of time and researching to get the device back to work, but happy to report it wasn't all for nothing!
